include/swift/SIL/SILBuilder.h

```diff
@@ -2058,54 +2058,121 @@ class SILBuilder {
+/// scope. It provides constructors which support inheriting a debug scope
+/// from the instruction being expanded, and constructors of convenience which
+/// infer the correct debug scope based on the insertion point.
+class SILBuilderForCodeExpansion : public SILBuilder {
+  using super = SILBuilder;
+
+  void inheritScopeFrom(SILInstruction &I) {
+    assert(I.getDebugScope() && "instruction has no debug scope");
+    setCurrentDebugScope(I.getDebugScope());
   }
 
 public:
-  explicit SILBuilderWithScope(
+  /// Creates a new SILBuilder with an insertion point before I. If
+  /// InsertedInstrs is specified, instructions emitted by the builder are
+  /// stored into it.
+  explicit SILBuilderForCodeExpansion(
       SILInstruction *I,
       SmallVectorImpl<SILInstruction *> *InsertedInstrs = nullptr)
       : SILBuilder(I, InsertedInstrs) {
-    assert(I->getDebugScope() && "instruction has no debug scope");
-    setCurrentDebugScope(I->getDebugScope());
+    inheritScopeFrom(*I);
   }
 
-  explicit SILBuilderWithScope(SILBasicBlock::iterator I)
-      : SILBuilderWithScope(&*I) {}
+  explicit SILBuilderForCodeExpansion(SILBasicBlock::iterator I)
+      : SILBuilderForCodeExpansion(&*I) {}
 
-  explicit SILBuilderWithScope(SILInstruction *I,
-                               SILInstruction *InheritScopeFrom)
-      : SILBuilderWithScope(I) {
-    inheritScopeFrom(InheritScopeFrom);
+  explicit SILBuilderForCodeExpansion(SILInstruction *I,
+                                      SILInstruction *InheritScopeFrom)
+      : SILBuilderForCodeExpansion(I) {
+    inheritScopeFrom(*InheritScopeFrom);
   }
 
-  explicit SILBuilderWithScope(SILBasicBlock::iterator I,
-                               SILInstruction *InheritScopeFrom)
-      : SILBuilderWithScope(&*I) {
-    inheritScopeFrom(InheritScopeFrom);
+  explicit SILBuilderForCodeExpansion(SILBasicBlock::iterator I,
+                                      SILInstruction *InheritScopeFrom)
+      : SILBuilder(I) {
+    inheritScopeFrom(*InheritScopeFrom);
   }
 
-  explicit SILBuilderWithScope(SILBasicBlock *BB,
-                               SILInstruction *InheritScopeFrom)
+  /// Creates a new SILBuilder with an insertion point at the end of BB.
+  explicit SILBuilderForCodeExpansion(SILBasicBlock *BB,
+                                      SILInstruction *InheritScopeFrom)
       : SILBuilder(BB) {
-    inheritScopeFrom(InheritScopeFrom);
+    inheritScopeFrom(*InheritScopeFrom);
+  }
+
+  /// Creates a new SILBuilder with an insertion point at the beginning of BB.
+  static SILBuilderForCodeExpansion
+  atBeginning(SILBasicBlock *BB, SILInstruction *InheritScopeFrom) {
+    assert(!BB->empty() && "No code to expand in basic block");
+    return SILBuilderForCodeExpansion(BB->begin(), InheritScopeFrom);
+  }
+  static SILBuilderForCodeExpansion atBeginning(SILBasicBlock *BB) {
+    assert(!BB->empty() && "No code to expand in basic block");
+    return atBeginning(BB, &*BB->begin());
```

Comment:
I think it would still be possible for getScopeOfFirstNonMetaInstruction() to pick up the wrong scope to apply. It might be better to require clients to be more explicit about which scope they want here.